STATISTICS-80: Variance Implementation

ani5rudh opened the pull request.

ani5rudh committed:
STATISTICS-79: Mean Implementation

ani5rudh committed:
STATISTICS-79: Mean Implementation

ani5rudh committed:
STATISTICS-79: Mean Implementation

ani5rudh committed:
Merge branch 'master' of https://github.com/apache/commons-statistics into ani5rudh/STATISTICS-79

ani5rudh committed:
STATISTICS-80: Variance Implementation

ani5rudh committed:
STATISTICS-80: Variance Implementation

ani5rudh committed:
STATISTICS-80: Variance Implementation

ani5rudh committed:
STATISTICS-80: Variance Implementation

kinow commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -0,0 +1,233 @@
+ * American Statistician, vol. 37, no. 3 (1983) pp. 242-247.</li></ul>
+ *
+ * Note that adding values using {@code accept} and then executing {@code getAsDouble} will
+ * sometimes give a different, less accurate, result than executing
```
Are we able to say anything more about accuracy? Maybe the papers linked above have something we could add here?

At $work we are using @tdunning's tdigest (a Python port based on it) for one-pass algorithms, where researchers are validating the computation and accuracy of statistics. They used both the paper and also notes from the javadocs that talk about the performance/accuracy tradeoff in the algorithm. e.g. https://github.com/tdunning/t-digest/blob/72acae1e14962795888efc941516c2b4503a86af/core/src/main/java/com/tdunning/math/stats/TDigest.java#L34 & https://github.com/tdunning/t-digest/blob/72acae1e14962795888efc941516c2b4503a86af/core/src/main/java/com/tdunning/math/stats/ScaleFunction.java#L21

Maybe we don't need to be very specific here, and we can just expand  a bit more and quote the article authors?

kinow replied:
BTW, this can be a follow-up issue, no need to block this PR for that, nor the other tasks needed for GSoC, I think.

ani5rudh replied:
I'll go through the papers once again and see if I can add some more information regarding accuracy. 👍

kinow commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -0,0 +1,233 @@
+        } else {
+            squaredDevSum = accum - (accum2 * accum2 / n);
+        }
+        return StorelessSampleVariance.create(squaredDevSum, mean, n, accum2 + (mean * n));
```
I started reading the paper but ran out of time to verify the code here, but from the javadocs I think it is based on this one?

![image](https://github.com/apache/commons-statistics/assets/304786/1650255a-2e62-47b1-8772-3cd3abca6f69)

ani5rudh replied:
West's algorithm is used in the `accept(double d)` method for updating the value of the statistic.
The two-pass algorithm used in the `of(double... values)` method is :

![image](https://github.com/apache/commons-statistics/assets/129569933/f1c9eb96-2926-4a5d-a61b-c14a5718c78a)
This formula is from the paper/report titled - [Algorithms for computing the sample variance : analysis and recommendations](https://www.cs.yale.edu/publications/techreports/tr222.pdf)

kinow replied:
Thanks for the explanation. I had a look at `SumOfSquaredDeviations#accept` and it looks correct to me!

kinow reviewed (commented):
Didn't have time to go through the tests, leaving a few comments, but found no blockers so far. Great work @ani5rudh ! :clap:

ani5rudh commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/SumOfSquaredDeviations.java:
```diff
@@ -0,0 +1,127 @@
+ * <p><strong>Note that this implementation is not synchronized.</strong> If
+ * multiple threads access an instance of this class concurrently, and at least
+ * one of the threads invokes the <code>increment()</code> or
+ * <code>clear()</code> method, it must be synchronized externally.
```
Thanks for pointing this @kinow. While using the javadocs from CM, I had forgotten to make the required changes. It should have been `accept()` and `combine()` instead of `increment()` and `clear()`. I'll update it in the next commit. Apologies for this silly error.

kinow replied:
Not a problem @ani5rudh. Probably worth vhecking other files in this PR and maybe the others previously written to confirm it doesn't jave the same text.

ani5rudh committed:
Update javadocs.

ani5rudh committed:
Update javadocs.

ani5rudh committed:
STATISTICS-80: Variance Implementation

ani5rudh committed:
Update javadocs.

ani5rudh committed:
Update javadocs.

ani5rudh committed:
Merge remote-tracking branch 'origin/ani5rudh/STATISTICS-80' into ani5rudh/STATISTICS-80

kinow commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -0,0 +1,233 @@
+        if (accum2Squared == Double.POSITIVE_INFINITY) {
+            squaredDevSum = Double.POSITIVE_INFINITY;
+        } else {
+            squaredDevSum = accum - (accum2 * accum2 / n);
```
I think here `(accum2 * accum2 / n)` could be simplified to `(accum2Squared / n)`.

kinow replied:
Had a look at the part of the paper you highlighted, @ani5rudh , and this looks correct to me! Great work!

kinow reviewed (commented):
Looks good to me. Reviewed implemented code and javadocs. Executed tests on my IDE, coverage looks great, couldn't find anything to suggest changing in the tests. Left a comment about docs for accuracy, but that's it from me. Thanks @ani5rudh !

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -0,0 +1,233 @@
+            if (n == 0) {
+                return Double.NaN;
+            } else if (n == 1 && Double.isFinite(sumOfSquaredDev)) {
+                return 0d;
```
No reason to use 0d here. This is only relevant when you wish for the zero to be used as double for double arithmetic, e.g. `1 / 0 != 1 / 0d` (the LHS will throw an arithmetic exception).

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -0,0 +1,233 @@
+            squaredDeviationSum.combine(that.squaredDeviationSum);
+            return this;
+        }
+
```
Remove empty line

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/MeanTest.java:
```diff
@@ -310,13 +308,4 @@ static Stream<Arguments> testCombineNonFinite() {
     }
-
-    // Helper function to compute the expected value of Mean using BigDecimal.
-    private static double computeExpected(double[] values) {
```
You could leave this function here and have the ported function return the BigDecimal result. This then becomes:
```java
private static double computeExpected(double[] values) {
    return TestHelper.computedExpectedMean(values).doubleValue();
}
```

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/VarianceTest.java:
```diff
@@ -0,0 +1,320 @@
+        for (double value : values) {
+            var.accept(value);
+        }
+        TestHelper.assertEquals(expected, var.getAsDouble(), 7, () -> "variance");
```
These ULP values may have to be lowered when you correct the computation of the expected variance. Using BigDecimal.valueOf(double) may have introduced quite a difference.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/VarianceTest.java:
```diff
@@ -0,0 +1,320 @@
+        }
+    }
+
+    static Stream<Arguments> testVariance() {
```
It may be better here to use MeanTest (since the data is the same):
```java
    static Stream<Arguments> testVariance() {
        return MeanTest.testMean();
    }
```

My preferred option would be to move the stream method to a new class: `TestData` with `static Stream<double[]> testValues()`. Note that you do not have to use JUnit `Arguments` when the test method only has one argument. You can simply stream the objects. JUnit will map the stream elements to the test argument.

Note: When I made the above change I also had the extra test case for Mean included. That fails:
```
Double.MAX_VALUE, Double.MAX_VALUE, -Double.MAX_VALUE

expected: Infinity
actual: NaN
```

This highlights that we really should reuse the same data on all the statistics, and it should be in one place.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/VarianceTest.java:
```diff
@@ -0,0 +1,320 @@
+        TestHelper.assertEquals(expected, actual, 2, () -> "array of arrays combined variance");
+    }
+
+    static Stream<Arguments> testCombine() {
```
Move this data to `TestData.testCombineValues`. This will have to use a Stream of Arguments.

aherbert reviewed (changes requested):
I think there is some work to do on consolidating the same test data that is used here and in the MeanTest. The stream of data can use the raw data. Each Test class can then derive an expected result. For the test data that includes non-finite values then I think we can simply use DoubleStream for the mean and use NaN for the variance.

I found a bug in the Mean implementation which we should fix outside of this PR.

ani5rudh committed:
- Refactor common testcases to a new class TestData.
- Refactor accept and combine methods.

ani5rudh committed:
- Refactor common testcases to a new class TestData.
- Refactor accept and combine methods.

ani5rudh committed:
Merge branch 'master' of https://github.com/apache/commons-statistics into ani5rudh/STATISTICS-80

ani5rudh committed:
Merge remote-tracking branch 'origin/ani5rudh/STATISTICS-80' into ani5rudh/STATISTICS-80

# Conflicts:
#	commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/SumOfSquaredDeviations.java
#	commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/MeanTest.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestHelper.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/VarianceTest.java

kinow commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestHelper.java:
```diff
@@ -48,14 +48,14 @@ static double[] concatenate(double[]... arrays) {
      * Helper function to compute the expected value of Mean using BigDecimal.
      * @param values Values.
-     * @return Mean of values.
+     * @return Mean of values rounded to <a href = "https://en.wikipedia.org/wiki/Decimal128_floating-point_format"> DECIMAL128 precision.
```
The `<a>` tag looks malformed to me, but if that renders valid Javadoc & HTML then I guess that's OK.

aherbert replied:
It opens a webpage in my IDE so the href is OK. But it should have a `</a>` tag. Note that testing with:
```
mvn javadoc:test-javadoc
```
errors as there are no public or protected test classes. So this link is just for internal use only.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/VarianceTest.java:
```diff
@@ -0,0 +1,259 @@
+        Assertions.assertEquals(expected, actual, "array of arrays combined variance non-finite");
+    }
+
+    static Stream<Arguments> testCombineVarianceNonFinite() {
```
Again, these are not the same as the MeanTest. So we are missing cases in one or the other. Please consolidate them to TestData.

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/SumOfSquaredDeviations.java:
```diff
@@ -0,0 +1,121 @@
+            final double sqDiffOfMean = diffOfMean * diffOfMean;
+            squaredDevSum += other.squaredDevSum + sqDiffOfMean * ((double) (oldN * otherN) / (oldN + otherN));
+        }
+        super.combine(other.getFirstMoment());
```
No need to create a copy instance here. Just use: `super.combine(other)`

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/SumOfSquaredDeviations.java:
```diff
@@ -0,0 +1,121 @@
+        if (oldN == 0) {
+            squaredDevSum = other.squaredDevSum;
+        } else if (otherN != 0) {
+            final double diffOfMean = other.getFirstMoment().getAsDouble() - m1;
```
Avoid creating a copy instance of the FirstMoment. Instead drop the implementation of the statistic interfaces (these are internal classes so do not require them):
```java
// In FirstMoment:
    double getFirstMoment() {
        if (Double.isFinite(m1)) {
            return n == 0 ? Double.NaN : m1;
        }
        // A non-finite value must have been encountered, return nonFiniteValue which represents m1.
        return nonFiniteValue;
    }

// In SumOfSquaredDeviations:
    final double diffOfMean = other.getFirstMoment() - m1
```

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/SumOfSquaredDeviations.java:
```diff
@@ -0,0 +1,121 @@
+ * provides the necessary partitioning, isolation, and merging of results for
+ * safe and efficient parallel execution.
+ */
+class SumOfSquaredDeviations extends FirstMoment implements DoubleStatistic {
```
I think we should drop the implementation of DoubleStatistic. You could just implement DoubleConsumer in FirstMoment and drop the implementation of DoubleStaticticAccumilator<FirstMoment>. You then do not have to overload getAsDouble through the hierarchy. You can have typed methods for combine  and specific methods for the stat:

```
FirstMoment implements DoubleConsumer:
void accept(double)
FirstMoment combine(FirstMoment);
double getFirstMoment()

SumOfSquaredDeviations extends FirstMoment:
SumOfSquaredDeviations combine(SumOfSquareDeviations);
double getSumOfSquaredDeviations();
```

This allows the SumOfSquareDeviations to call `super.combine(other)` and `getFirstMoment()` directly without creating copy instances.

aherbert reviewed (changes requested):
I think we can simplify the inheritance of the internal classes by dropping the interfaces implementation. We still keep the methods, but without the interface definition we can have separate methods for each class so the SumOfSquaredDeviations can compute the mean.

There are some other changes to consolidate the test cases in the TestData class.

aherbert committed:
Add the observed ULP difference to the assert message

ani5rudh committed:
Consolidate test cases for non-finites.
- Refactor of() method to better handle overflow.
- Drop the implementation of the statistic and accumulator interfaces.

ani5rudh committed:
Reduce the ULP tolerance.

ani5rudh committed:
Merge remote-tracking branch 'origin/ani5rudh/STATISTICS-80' into ani5rudh/STATISTICS-80

# Conflicts:
#	commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/SumOfSquaredDeviations.java
#	commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/MeanTest.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestData.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestHelper.java
#	commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/VarianceTest.java

ani5rudh committed:
Remove unused methods.

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -109,6 +109,9 @@ public static Variance create() {
     public static Variance of(double... values) {
         final double mean = Mean.of(values).getAsDouble();
+        if (!Double.isFinite(mean)) {
+            return StorelessSampleVariance.create(Math.abs(mean), mean, values.length, Math.abs(mean));
```
Correction
```java
return StorelessSampleVariance.create(Math.abs(mean), mean, values.length, mean);
```
The second value is a proxy for the non-finite sum of the values, so we should not use abs.

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -124,10 +127,10 @@ public static Variance of(double... values) {
         // To prevent squaredDevSum from spuriously attaining a NaN value
         // when accum2Squared (which implies accum is also infinite) is infinite, assign it
```
This comment is now out-of-date.

You should also update the method javadoc with the case: if the sum of the squared deviations from the mean is infinite, the variance is NaN.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestData.java:
```diff
@@ -19,11 +19,18 @@
     private TestData() {}
 
+    /**
+     * Function which supplies data to test the <code>accept()</code> and <code>of()</code> methods.
```
supplies a test data for a statistic in a single array.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestData.java:
```diff
@@ -60,18 +67,33 @@ static Stream<double[]> testValues() {
 
-    static Stream<Arguments> testValuesNonFinite() {
+    /**
+     * Function which supplies data with non-finite values to test the <code>accept()</code> and <code>of()</code> methods.
```
supplies a test data for a statistic in a single array. Each case will contain at least 1 non-finite value.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestData.java:
```diff
@@ -60,18 +67,33 @@ static Stream<double[]> testValues() {
     }
 
+    /**
+     * Function which supplies data to test the <code>combine()</code> method.
```
supplies a test data for a statistic in as a pair of double[] arrays.

aherbert commented on commons-statistics-descriptive/src/test/java/org/apache/commons/statistics/descriptive/TestData.java:
```diff
@@ -100,15 +122,26 @@ static Stream<Arguments> testCombine() {
 
-    static Stream<Arguments> testCombineNonFinite() {
+    /**
+     * Function which supplies data with non-finite values to test the <code>combine()</code> method.
```
supplies a test data for a statistic in as a pair of double[] arrays. Each case will contain at least 1 non-finite value.

aherbert reviewed (changes requested):
This is looking good. I've made a few suggestions to change the description of the TestData methods to be more generic. You have referenced the `accept` and `of` methods but the data could be used to test other statistic methods. So we should make the intended purpose of the data more clear.

ani5rudh committed:
Update javadocs.
- Better handle overflow in combine() method.

aherbert commented on commons-statistics-descriptive/src/main/java/org/apache/commons/statistics/descriptive/Variance.java:
```diff
@@ -97,6 +97,7 @@ public static Variance create() {
      *     <li>the input array is empty,</li>
      *     <li>any of the values is <code>NaN</code>, or</li>
      *     <li>an infinite value of either sign is encountered</li>
+     *     <li>if the sum of the squared deviations from the mean is infinite</li>
```
Your `or` is now on the wrong line

ani5rudh committed:
Update javadocs.

ani5rudh committed:
Update javadocs.
